master/docs/developer/stats-service.rst

```diff
@@ -0,0 +1,255 @@
+(*Note*: This machine may be different from the one running :class:`BuildMaster`)
+
+Data is captured according to the master config file and then, is sent to each of the storage backends provided by the master configuration (see :bb:cfg:`stats-service`).
+
+Each storage backend has a Python client defined as part of :mod:`buildbot.statistics.storage_backends`
+to aid in posting data by :class:`StatsService`
+
+Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage backend.
+
+.. py:class:: StatsStorageBase
+
+   A base class for all storage services
+
+.. py:class:: InfluxStorageService
+
+   `InfluxDB <http://influxdb.com>`_ is a distributed, time series database that employs a key-value pair storage system.
+
+   This class is a Buildbot client to the InfluxDB storage backend.
+   It is available in the configuration as ``stats.InfluxStorageService``
+   It takes the following initialization arguments:
+
+   * ``url``: The URL where the service is running.
+   * ``port``: The port on which the service is listening.
+   * ``user``: Username of a InfluxDB user.
+   * ``password``: Password for ``user``.
+   * ``db``: The name of database to be used.
+   * ``captures``: A list of instances of subclasses of :py:class:`Capture`. This tells which stats are to be stored in this storage backend.
+   * ``name=None``: (Optional) The name of this storage backend.
+
+   .. py:method:: postStatsValue(self, post_data, series_name, context={})
+
+      :param post_data: A dict of key-value pairs that is sent for storage. The keys of this dict can be thought of as columns in a database and the value is the data stored for that column.
+      :type post_data: dict
+      :param series_name: The name of the time-series for this statistic.
+      :type series_name: str
+      :param context: (Optional) Any other contextual information about the data. Dict of key-value pairs.
+      :type context: dict
+
+      This method constructs a dictionary of data to be sent to InfluxDB in the proper format and sends the data to the influxDB instance.
+
+
+Capture Classes
+---------------
+
+Capture classes are used for declaring the data that needs to be sent to storage backends for storage.
+
+.. py:class:: Capture
+
+   A base class for all capture classes. Not to be used directly. Initlized with the following parameters:
+
+   * ``routingKey``: (tuple) The routing key to be used by :class:`StatsService` to register consumers to the MQ lasyer for the subclass of this class.
+   * ``callback``: The callback registered with the MQ layer for the consumer of a subclass of this class. Each subclass must provide a default callback for this purpose.
+
+   .. py:method:: defaultContext(self, msg):
+
+      A method for providing default context to the storage backends.
+
+   .. py:method:: consumer(self, routingKey, msg):
+
+      Raises ``NotImplementedError``. Each subclass of this method should implement its own consumer.
```

Comment:
Instead of "Raises `NotImplementedError`", I'd suggest to write "each subclass of class `Capture` must implement this method, which is expected to do ..."